### PR TITLE
[ARPACK] Update to 3.9.1

### PR DESCRIPTION
**Pull request by ViralBShah**

### CODE REVIEW

**Comment by ViralBShah** [comment]
@giordano I haven't seen this error on Yggdrasil before. Any idea how one might fix it?

```
2024-01-16 11:40:05 UTC | [11:40:05] CMake Error at /usr/share/cmake/Modules/FortranCInterface.cmake:391 (message):
-- | --
  | 2024-01-16 11:40:05 UTC | [11:40:05]   The Fortran compiler:
  | 2024-01-16 11:40:05 UTC | [11:40:05]
  | 2024-01-16 11:40:05 UTC | [11:40:05]     /opt/bin/x86_64-apple-darwin14-libgfortran4-cxx11/x86_64-apple-darwin14-gfortran
  | 2024-01-16 11:40:05 UTC | [11:40:05]
  | 2024-01-16 11:40:05 UTC | [11:40:05]   and the C compiler:
  | 2024-01-16 11:40:05 UTC | [11:40:05]
  | 2024-01-16 11:40:05 UTC | [11:40:05]     /opt/bin/x86_64-apple-darwin14-libgfortran4-cxx11/x86_64-apple-darwin14-clang
  | 2024-01-16 11:40:05 UTC | [11:40:05]
  | 2024-01-16 11:40:05 UTC | [11:40:05]   failed to compile a simple test project using both languages.  The output
  | 2024-01-16 11:40:05 UTC | [11:40:05]   was:
  | 2024-01-16 11:40:05 UTC | [11:40:05]
  | 2024-01-16 11:40:05 UTC | [11:40:05]     Change Dir: /workspace/srcdir/arpack-ng/build/CMakeFiles/FortranCInterface/VerifyC
  | 2024-01-16 11:40:05 UTC | [11:40:05]
```

**Comment by giordano** [comment]
Impossible to tell without reading the logs: https://docs.binarybuilder.org/stable/troubleshooting/#General-comments

**Comment by ViralBShah** [comment]
The log is actually in there, helpfully. The issue is:

```
2024-01-16 06:40:05 EST | [11:40:05]     /opt/bin/x86_64-apple-darwin14-libgfortran4-cxx11/x86_64-apple-darwin14-clang --sysroot=/opt/x86_64-apple-darwin14/x86_64-apple-darwin14/sys-root  -Dsnaupd=snaupd_64 -Dsneupd=sneupd_64 -Ddnaupd=dnaupd_64 -Ddneupd=dneupd_64 -DINCLUDE_FCMANGLE -O3 -DNDEBUG -mmacosx-version-min=10.10 -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/VerifyFortranC.dir/main.c.o CMakeFiles/VerifyFortranC.dir/VerifyC.c.o -o VerifyFortranC   -L/opt/x86_64-apple-darwin14/lib/gcc/x86_64-apple-darwin14/7.1.0  -L/opt/x86_64-apple-darwin14/lib/gcc  libVerifyFortran.a -lgfortran -lgcc_ext.10.5 -lgcc -lquadmath -lm -lgcc_ext.10.5 -lgcc
-- | --
  | 2024-01-16 06:40:05 EST | [11:40:05]     ld64.lld: error: No LC_DYLD_INFO_ONLY or LC_DYLD_EXPORTS_TRIE found in /opt/x86_64-apple-darwin14/x86_64-apple-darwin14/lib/libgcc_ext.10.5.dylib
  | 2024-01-16 06:40:05 EST | [11:40:05]     clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```

File: A/Arpack/build_tarballs.jl [hunk 1]
```diff
@@ -2,102 +2,46 @@ using BinaryBuilder
...
+        -DBUILD_SHARED_LIBS=ON \
+        -DINTERFACE64=1 -DSYMBOLSUFFIX=_64
+else
+    cmake .. -DCMAKE_INSTALL_PREFIX="$prefix" \
+        -DCMAKE_TOOLCHAIN_FILE="${CMAKE_TARGET_TOOLCHAIN}" -DCMAKE_BUILD_TYPE=Release \
+        -DBLAS_LIBRARIES=${LBT} -DBLAS_LINKER_FLAGS="-L${libdir}" \
+        -DLAPACK_LIBRARIES=${LBT} -DLAPACK_LINKER_FLAGS="-L${libdir}" \
+        -DEXAMPLES=OFF \
+        -DBUILD_SHARED_LIBS=ON
 fi
```
Comment (giordano) [inline A/Arpack/build_tarballs.jl:40]: This greatly reduces duplication
```suggestion
if [[ ${nbits} == 64 ]]; then
    FLAGS=(-DINTERFACE64=1 -DSYMBOLSUFFIX=_64)
fi
cmake .. -DCMAKE_INSTALL_PREFIX="$prefix" \
    -DCMAKE_TOOLCHAIN_FILE="${CMAKE_TARGET_TOOLCHAIN}" -DCMAKE_BUILD_TYPE=Release \
    -DBLAS_LIBRARIES=${LBT} -DBLAS_LINKER_FLAGS="-L${libdir}" \
    -DLAPACK_LIBRARIES=${LBT} -DLAPACK_LINKER_FLAGS="-L${libdir}" \
    -DEXAMPLES=OFF \
    -DBUILD_SHARED_LIBS=ON \
    "${FLAGS[@]}"
```

File: A/Arpack/build_tarballs.jl [hunk 2]
```diff
@@ -123,4 +67,4 @@ dependencies = [
 ]
 
 # Build the tarballs, and possibly a `build.jl` as well.
-build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version=v"6", julia_compat="1.8")
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version=v"8", julia_compat="1.8")
```
Comment (giordano) [inline A/Arpack/build_tarballs.jl:70]: > The log is actually in there, helpfully. The issue is:

Try not using lld to link, sometimes it's still buggy:
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; preferred_gcc_version=v"6", julia_compat="1.8", clang_use_lld=false)
```

File: A/Arpack/build_tarballs.jl (hunk 1 above)
Comment (ViralBShah) [reply A/Arpack/build_tarballs.jl:40]: Thanks. I was just being lazy since I couldn't remember how to do it.

**Comment by ViralBShah** [comment]
There's an upstream issue that needs addressing: https://github.com/opencollab/arpack-ng/pull/32#issuecomment-1894448658

**Pull request merged**